### PR TITLE
vpr: Convert 2D nested vectors storing net-pin data to a (sparse) flat matrix

### PR DESCRIPTION
**Pull request by kmurray**
vtr::FlatSparseMatrix takes into account that some rows (nets) are
smaller than others and lays them out in adjacent memory. This should
reduce memory fragmentation and improve locality.

### CODE REVIEW

File: vpr/src/route/route_budgets.cpp [hunk 1]
```diff
@@ -66,11 +66,11 @@ void route_budgets::free_budgets() {
     /*Free associated budget memory if set
      * if not set, only free the chunk memory that wasn't used*/
     if (set) {
-        free_net_delay(delay_min_budget, &min_budget_delay_ch);
-        free_net_delay(delay_max_budget, &max_budget_delay_ch);
-        free_net_delay(delay_target, &target_budget_delay_ch);
-        free_net_delay(delay_lower_bound, &lower_bound_delay_ch);
-        free_net_delay(delay_upper_bound, &upper_bound_delay_ch);
+        delay_min_budget.clear();
```
Comment (litghost) [inline vpr/src/route/route_budgets.cpp:69]: clear does not free the memory, only sets the size to 0.  Is this the intention?.  shrink_to_fit increases the chance that the memory is free'd, and the capacity is set to 0.

File: libs/libvtrutil/src/vtr_ragged_matrix.h [hunk 2]
```diff
@@ -0,0 +1,188 @@
+#ifndef VTR_RAGGED_MATRIX_H
```
Comment (litghost) [inline libs/libvtrutil/src/vtr_ragged_matrix.h:1]: This needs unit tests in libs/libvtrutils/test

File: libs/libvtrutil/src/vtr_ragged_matrix.h [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+      private:
+        size_t irow_;
+        Callback& callback_;
+    };
+
+  private:
+    std::vector<T> data_;
+    std::vector<int> first_elem_;
```
Comment (litghost) [inline libs/libvtrutil/src/vtr_ragged_matrix.h:183]: Consider make a method to sanity check that first_elem_ is sorted?

File: vpr/src/route/route_budgets.cpp (hunk 1 above)
Comment (kmurray) [reply vpr/src/route/route_budgets.cpp]: Good point, I've added vtr::release_memory(), which will force container memory to be released.

File: libs/libvtrutil/src/vtr_ragged_matrix.h (hunk 2 above)
Comment (kmurray) [reply libs/libvtrutil/src/vtr_ragged_matrix.h:1]: Done

File: libs/libvtrutil/src/vtr_ragged_matrix.h (hunk 3 above)
Comment (kmurray) [reply libs/libvtrutil/src/vtr_ragged_matrix.h:218]: The structure is enforced by the constructor and there is no method to modify it after construction, so no need to check for sorting.

File: libs/libvtrutil/src/vtr_ragged_matrix.h [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    friend void swap(FlatRaggedMatrix<T, Index0, Index1>& lhs, FlatRaggedMatrix<T, Index0, Index1>& rhs) {
+        lhs.swap(rhs);
+    }
+
+  public: //Types
+    //Proxy class used to represent a 'row' in the matrix
+    template<typename U>
+    class ProxyRow {
```
Comment (litghost) [inline libs/libvtrutil/src/vtr_ragged_matrix.h:142]: Can you just use array_view here, rather than defining a new thing?

Comment (kmurray) [reply libs/libvtrutil/src/vtr_ragged_matrix.h:143]: Yep, makes sense. I've updated it.

**Pull request merged**